Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/ReflectometryISISCalibration.py

```diff
@@ -0,0 +1,151 @@
+
+        return specular_pixel_idx
+
+    def _get_pixel_positions_from_file(self):
+        # Create dictionary of pixel spectrum number and position from scan
+        scanned_pixel_positions = {}
+        with open(self.getPropertyValue(self._CALIBRATION_FILE), 'r') as file:
+            pixels = csv.reader(file)
+            for pixel in pixels:
+                pixel_info = pixel[0].split()
+                scanned_pixel_positions[int(pixel_info[0])] = float(pixel_info[1])
+        return scanned_pixel_positions
+
+    def _create_calibration_table_from_scan(self, ws, det_info, comp_info, specular_pixel_idx):
+        scanned_pixel_positions = self._get_pixel_positions_from_file()
+
+        # Get the current specular pixel Y position and the position of the specular pixel from the calibration scan
+        specpixel_y = ws.getDetector(specular_pixel_idx).getPos().Y()
+
+        scanned_specpixel_y = scanned_pixel_positions.get(specular_pixel_idx)
+        if not scanned_specpixel_y:
+            raise RuntimeError("Could not find calibration data for specular pixel")
+
+        table = CreateEmptyTableWorkspace(OutputWorkspace="CalibTable")
+        table.addColumn(type="int", name="Detector ID")
+        table.addColumn(type="V3D", name="Detector Position")
+        table.addColumn(type="double", name="Detector Y Coordinate")
+        table.addColumn(type="double", name="Detector Height")
+        table.addColumn(type="double", name="Detector Width")
+
+        # Populate the table with the calibrated data
+        for ws_id, scanned_pixel_pos in scanned_pixel_positions.items():
+            try:
+                det = ws.getDetector(ws_id)
+            except RuntimeError:
+                raise RuntimeError(f"Could not find detector to calibrate at workspace id {ws_id}")
+
+            det_idx = det_info.indexOf(det.getID())  # detectorInfo and componentInfo index
+            box = comp_info.shape(det_idx).getBoundingBox().width()
+            scalings = comp_info.scaleFactor(det_idx)
+
+            next_row = {'Detector ID': det.getID(),
+                        'Detector Position': det.getPos(),
+                        'Detector Y Coordinate': self._calculate_calibrated_y_pos(specpixel_y, scanned_specpixel_y,
+                                                                                  scanned_pixel_pos),
+                        'Detector Width': box[0] * scalings[0],
+                        'Detector Height': box[1] * scalings[1]}
+            table.addRow(next_row)
+        return table
+
+    def _calculate_calibrated_y_pos(self, specpixel_y, scanned_specpixel_y, scanned_pixel_pos):
+        # First find the difference between the specular pixel position and the pixel position in the scan data.
+        # The calibrated Y co-ordinate is then found by adding this difference to the specular pixel position from
+        # the IDF.
+        # Need to confirm what the multiply 0.001 is for.
```

Comment:
I think the input file must be in `mm` so this must be the conversion to `m`. It's probably worth mentioning the expected units in the doc string for the input file.

Reply:
Ah, thanks, that's a good spot! I'll update the doc/comments.